File: majorProject/src/main/java/com/aocc/majorproject/ui/UiText.java
package com.aocc.majorproject.ui;

import android.graphics.Paint;

import com.aocc.framework.Graphics;

/** Text rendering with correct horizontal and vertical alignment within bounds. */
public final class UiText {

    public enum HAlign {
        LEFT(Paint.Align.LEFT),
        CENTER(Paint.Align.CENTER),
        RIGHT(Paint.Align.RIGHT);

        private final Paint.Align paintAlign;

        HAlign(Paint.Align paintAlign) {
            this.paintAlign = paintAlign;
        }
    }

    private UiText() {
    }

    public static void drawInBounds(Graphics g, Paint paint, String text, UiBounds bounds,
            HAlign hAlign, int color) {
        paint.setTextAlign(hAlign.paintAlign);
        int anchorX;
        switch (hAlign) {
            case LEFT:
                anchorX = bounds.x;
                break;
            case RIGHT:
                anchorX = bounds.x + bounds.width;
                break;
            default:
                anchorX = bounds.centerX();
                break;
        }
        drawAtBaseline(g, paint, text, anchorX, bounds.centerY(), color);
    }

    public static void drawCentered(Graphics g, Paint paint, String text, int centerX, int centerY,
            int color) {
        paint.setTextAlign(Paint.Align.CENTER);
        drawAtBaseline(g, paint, text, centerX, centerY, color);
    }

    public static void drawLeftOfCenter(Graphics g, Paint paint, String text, int leftX, int centerY,
            int color) {
        paint.setTextAlign(Paint.Align.LEFT);
        drawAtBaseline(g, paint, text, leftX, centerY, color);
    }

    private static void drawAtBaseline(Graphics g, Paint paint, String text, int anchorX, int centerY,
            int color) {
        Paint.FontMetrics fm = paint.getFontMetrics();
        float baseline = centerY - (fm.ascent + fm.descent) / 2f;
        g.drawString(text, anchorX, Math.round(baseline), color, paint);
    }
}

File: majorProject/src/main/kotlin/com/aocc/majorproject/ui/UiText.kt
package com.aocc.majorproject.ui

import android.graphics.Paint
import com.aocc.framework.Graphics

/** Text rendering with correct horizontal and vertical alignment within bounds. */
object UiText {

    enum class HAlign(private val paintAlign: Paint.Align) {
        LEFT(Paint.Align.LEFT),
        CENTER(Paint.Align.CENTER),
        RIGHT(Paint.Align.RIGHT),
    }

    @JvmStatic
    fun drawInBounds(
        g: Graphics,
        paint: Paint,
        text: String,
        bounds: UiBounds,
        hAlign: HAlign,
        color: Int,
    ) {
        paint.textAlign = when (hAlign) {
            HAlign.LEFT -> Paint.Align.LEFT
            HAlign.RIGHT -> Paint.Align.RIGHT
            HAlign.CENTER -> Paint.Align.CENTER
        }
        val anchorX = when (hAlign) {
            HAlign.LEFT -> bounds.x
            HAlign.RIGHT -> bounds.x + bounds.width
            HAlign.CENTER -> bounds.centerX()
        }
        drawAtBaseline(g, paint, text, anchorX, bounds.centerY(), color)
    }

    @JvmStatic
    fun drawCentered(
        g: Graphics,
        paint: Paint,
        text: String,
        centerX: Int,
        centerY: Int,
        color: Int,
    ) {
        paint.textAlign = Paint.Align.CENTER
        drawAtBaseline(g, paint, text, centerX, centerY, color)
    }

    @JvmStatic
    fun drawLeftOfCenter(
        g: Graphics,
        paint: Paint,
        text: String,
        leftX: Int,
        centerY: Int,
        color: Int,
    ) {
        paint.textAlign = Paint.Align.LEFT
        drawAtBaseline(g, paint, text, leftX, centerY, color)
    }

    private fun drawAtBaseline(
        g: Graphics,
        paint: Paint,
        text: String,
        anchorX: Int,
        centerY: Int,
        color: Int,
    ) {
        val fm = paint.fontMetrics
        val baseline = centerY - (fm.ascent + fm.descent) / 2f
        g.drawString(text, anchorX, baseline.toInt(), color, paint)
    }
}
